File: packages/shared-types/src/incidente.ts
/**
 * 📁 packages/shared-types/src/incidente.ts
 * 🎯 Interfaces TypeScript del expediente único de incidente y contratos del API.
 * 📦 Capa: Packages / Shared Types
 *
 * Estas interfaces son el contrato compartido entre backend (FastAPI/Pydantic)
 * y frontends (web + mobile). Mantener en sync con app/schemas/incidente.py.
 */

import { EstadoIncidente, NivelSeveridad, TipoCanal } from "./enums";

export interface Incidente {
  id: string;
  codigo: string; // INC-YYYYMMDD-XXXXX
  titulo: string;
  descripcion: string | null;
  estado: EstadoIncidente;
  severidad: NivelSeveridad;
  categoria: string | null;
  canal_origen: TipoCanal;
  latitud: number | null;
  longitud: number | null;
  reportante_id: string;
  operador_asignado_id: string | null;
  supervisor_id: string | null;
  created_at: string;
  updated_at: string;
}

// --- API contracts ---------------------------------------------------------

/** Item devuelto por GET /api/v1/incidentes y GET /api/v1/incidentes/mis. */
export interface IncidenteListItem {
  id: string;
  codigo: string;
  titulo: string;
  descripcion: string | null;
  estado: EstadoIncidente;
  severidad: NivelSeveridad | null;
  categoria: string | null;
  lugar_referencia: string | null;
  latitud: number | null;
  longitud: number | null;
  canal_origen: TipoCanal;
  operador_nombre: string | null;
  operador_avatar_url: string | null;
  created_at: string;
}

/** Filtros aceptados por GET /api/v1/incidentes (vista operativa). */
export interface IncidenteListFilters {
  search?: string;
  severidad?: NivelSeveridad;
  estado?: EstadoIncidente;
  limit?: number;
}

/** Representación reducida de usuario en respuestas de incidente. */
export interface UsuarioMini {
  id: string;
  nombre_completo: string;
  email: string | null;
  avatar_url: string | null;
}

/** Evento del historial de un incidente. */
export interface HistorialEvento {
  id: string;
  estado_anterior: EstadoIncidente | null;
  estado_nuevo: EstadoIncidente;
  accion: string;
  comentario: string | null;
  ejecutado_por: UsuarioMini | null;
  created_at: string;
}

/** Mensaje asociado al expediente de incidente. */
export interface ComentarioIncidenteItem {
  id: string;
  incidente_id: string;
  autor: UsuarioMini | null;
  contenido: string;
  es_interno: boolean;
  created_at: string;
  updated_at: string;
}

/** Evidencia documental o multimedia asociada al expediente. */
export interface EvidenciaIncidenteItem {
  id: string;
  incidente_id: string;
  tipo_archivo: string;
  nombre_archivo: string;
  url_archivo: string;
  tamano_bytes: number | null;
  mime_type: string | null;
  descripcion: string | null;
  cargado_por: UsuarioMini | null;
  created_at: string;
}

/** Snapshot formal generado cuando el incidente se cierra. */
export interface ExpedienteCierre {
  id: string;
  incidente_id: string;
  resumen_cierre: string;
  resultado: string | null;
  snapshot: Record<string, unknown>;
  generado_por: UsuarioMini | null;
  pdf_url: string | null;
  created_at: string;
  updated_at: string;
}

/** Borrador generado por IA para completar el expediente de cierre. */
export interface ExpedienteCierreAiDraft {
  resumen_cierre: string;
  resultado_cierre: string | null;
}

/** Resultado de priorizacion automatica de un incidente. */
export interface IncidentePriorizacionAi {
  severidad: NivelSeveridad;
  categoria_sugerida: string | null;
  confianza: number | null;
  justificacion: string | null;
}

/** Conteo de incidentes por zona (lugar_referencia). */
export interface ZonaCount {
  zona: string;
  total: number;
}

/** Respuesta de GET /api/v1/incidentes/stats. */
export interface DashboardStats {
  total: number;
  activos: number;
  criticos: number;
  en_atencion: number;
  resueltos_24h: number;
  por_zona: ZonaCount[];
}

/** Métrica con valor actual y % de cambio respecto al periodo anterior. */
export interface KpiCard {
  valor: number;
  cambio_pct: number;
  unidad: string;
}

/** Punto de evolución diaria en el chart de timeseries. */
export interface EvolucionPunto {
  fecha: string; // YYYY-MM-DD
  total: number;
  resueltos: number;
  criticos: number;
}

/** Conteo por tipo (categoría) con porcentaje del total. */
export interface TipoCount {
  tipo: string;
  total: number;
  porcentaje: number;
}

/** Indicador SLA con valor actual vs objetivo. */
export interface SlaIndicador {
  actual: number;
  objetivo: number;
  unidad: string;
}

export type KpisPeriod = "semana" | "mes" | "trimestre" | "año";

/** Respuesta de GET /api/v1/incidentes/kpis. */
export interface KpisResponse {
  period: KpisPeriod;
  frt: KpiCard;
  tmr: KpiCard;
  total_incidentes: KpiCard;
  tasa_resolucion: KpiCard;
  criticos: KpiCard;
  sla_cumplimiento: KpiCard;
  evolucion: EvolucionPunto[];
  por_tipo: TipoCount[];
  por_zona: ZonaCount[];
  sla: {
    frt: SlaIndicador;
    tmr: SlaIndicador;
    escalamiento: SlaIndicador;
    criticos_sla: SlaIndicador;
  };
}

/** Respuesta de GET /api/v1/incidentes/{id}. */
export interface IncidenteDetail {
  id: string;
  codigo: string;
  titulo: string;
  descripcion: string | null;
  estado: EstadoIncidente;
  severidad: NivelSeveridad | null;
  categoria: string | null;
  lugar_referencia: string | null;
  latitud: number | null;
  longitud: number | null;
  canal_origen: TipoCanal;
  fecha_primera_respuesta: string | null;
  fecha_resolucion: string | null;
  created_at: string;
  updated_at: string;
  reportante: UsuarioMini | null;
  operador_asignado: UsuarioMini | null;
  supervisor: UsuarioMini | null;
  historial: HistorialEvento[];
  comentarios: ComentarioIncidenteItem[];
  evidencias: EvidenciaIncidenteItem[];
  expediente_cierre: ExpedienteCierre | null;
}

/** Wrapper de respuesta para listados. */
export interface IncidenteListResponse {
  items: IncidenteListItem[];
  total: number;
}

/** Item para mapa tactico operativo. */
export interface IncidenteMapaItem {
  id: string;
  codigo: string;
  titulo: string;
  estado: EstadoIncidente;
  severidad: NivelSeveridad | null;
  categoria: string | null;
  lugar_referencia: string | null;
  latitud: number | null;
  longitud: number | null;
  created_at: string | null;
}

/** Respuesta de GET /api/v1/incidentes/mapa. */
export interface IncidenteMapaResponse {
  items: IncidenteMapaItem[];
  total: number;
  georreferenciados: number;
  sin_coordenadas: number;
}

/** Body de POST /api/v1/incidentes (creación desde el PWA Comunidad). */
export interface IncidenteCreateInput {
  titulo: string;
  descripcion?: string | null;
  severidad?: NivelSeveridad | null;
  categoria?: string | null;
  lugar_referencia?: string | null;
  latitud?: number | null;
  longitud?: number | null;
}

/** Respuesta de POST /api/v1/incidentes. */
export interface IncidenteCreated {
  id: string;
  codigo: string;
  estado: EstadoIncidente;
  created_at: string;
}

/** Body de PATCH /api/v1/incidentes/{id}/estado. */
export interface IncidenteEstadoUpdate {
  estado: EstadoIncidente;
  comentario?: string | null;
  resumen_cierre?: string | null;
  resultado_cierre?: string | null;
}

/** Body de PATCH /api/v1/incidentes/{id}/asignar. */
export interface IncidenteAsignacionUpdate {
  operador_asignado_id: string;
  comentario?: string | null;
}

/** Body de POST /api/v1/incidentes/{id}/comentarios. */
export interface ComentarioIncidenteCreateInput {
  contenido: string;
  es_interno?: boolean;
}

/** Item devuelto por GET /api/v1/incidentes/operadores. */
export interface OperadorListItem {
  id: string;
  nombre_completo: string;
  email: string;
  avatar_url: string | null;
  rol: string;
}

File: apps/backend/app/schemas/incidente.py
"""
📁 apps/backend/app/schemas/incidente.py
🎯 Schemas Pydantic del módulo de incidentes — request/response del API.
📦 Capa: Schemas
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.core.constants import EstadoIncidente, NivelSeveridad, TipoCanal


class IncidenteListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    codigo: str
    titulo: str
    descripcion: str | None = None
    estado: EstadoIncidente
    severidad: NivelSeveridad | None = None
    categoria: str | None = None
    lugar_referencia: str | None = None
    latitud: float | None = None
    longitud: float | None = None
    live_location_enabled: bool = False
    live_location_updated_at: datetime | None = None
    live_location_expires_at: datetime | None = None
    canal_origen: TipoCanal
    operador_nombre: str | None = None
    operador_avatar_url: str | None = None
    created_at: datetime | None = None


class IncidenteListResponse(BaseModel):
    items: list[IncidenteListItem]
    total: int


class IncidenteMapaItem(BaseModel):
    id: str
    codigo: str
    titulo: str
    estado: EstadoIncidente
    severidad: NivelSeveridad | None = None
    categoria: str | None = None
    lugar_referencia: str | None = None
    latitud: float | None = None
    longitud: float | None = None
    live_location_enabled: bool = False
    live_location_updated_at: datetime | None = None
    created_at: datetime | None = None


class IncidenteMapaResponse(BaseModel):
    items: list[IncidenteMapaItem]
    total: int
    georreferenciados: int
    sin_coordenadas: int


class IncidenteCreateInput(BaseModel):
    titulo: str = Field(min_length=3, max_length=200)
    descripcion: str | None = Field(default=None, max_length=4000)
    severidad: NivelSeveridad | None = None
    categoria: str | None = Field(default=None, max_length=100)
    lugar_referencia: str | None = Field(default=None, max_length=255)
    latitud: float | None = None
    longitud: float | None = None
    canal_origen: TipoCanal = TipoCanal.WEB


class IncidenteCreated(BaseModel):
    id: str
    codigo: str
    estado: EstadoIncidente
    created_at: datetime


class IncidenteLiveLocationUpdate(BaseModel):
    latitud: float | None = Field(default=None, ge=-90, le=90)
    longitud: float | None = Field(default=None, ge=-180, le=180)
    precision_metros: float | None = Field(default=None, ge=0, le=10000)
    activo: bool = True


class IncidenteEstadoUpdate(BaseModel):
    estado: EstadoIncidente
    comentario: str | None = Field(default=None, max_length=2000)
    resumen_cierre: str | None = Field(default=None, min_length=20, max_length=6000)
    resultado_cierre: str | None = Field(default=None, max_length=2000)


class ExpedienteCierreAiDraft(BaseModel):
    resumen_cierre: str = Field(min_length=20, max_length=6000)
    resultado_cierre: str | None = Field(default=None, max_length=2000)


class IncidentePriorizacionAi(BaseModel):
    severidad: NivelSeveridad
    categoria_sugerida: str | None = Field(default=None, max_length=100)
    confianza: float | None = Field(default=None, ge=0, le=1)
    justificacion: str | None = Field(default=None, max_length=1200)


class IncidenteAsignacionUpdate(BaseModel):
    operador_asignado_id: str  # UUID
    comentario: str | None = Field(default=None, max_length=2000)


class OperadorListItem(BaseModel):
    id: str
    nombre_completo: str
    email: EmailStr
    avatar_url: str | None = None
    rol: str  # "operador" | "supervisor"


class UsuarioMini(BaseModel):
    """Representación reducida de un usuario para mostrar en el detalle."""

    id: str
    nombre_completo: str
    email: EmailStr | None = None
    avatar_url: str | None = None
    rol: str | None = None


class HistorialEvento(BaseModel):
    id: str
    estado_anterior: EstadoIncidente | None = None
    estado_nuevo: EstadoIncidente
    accion: str
    comentario: str | None = None
    ejecutado_por: UsuarioMini | None = None
    created_at: datetime


class ComentarioIncidenteItem(BaseModel):
    id: str
    incidente_id: str
    autor: UsuarioMini | None = None
    contenido: str
    es_interno: bool
    created_at: datetime
    updated_at: datetime


class EvidenciaIncidenteItem(BaseModel):
    id: str
    incidente_id: str
    tipo_archivo: str
    nombre_archivo: str
    url_archivo: str
    tamano_bytes: int | None = None
    mime_type: str | None = None
    descripcion: str | None = None
    cargado_por: UsuarioMini | None = None
    created_at: datetime


class ExpedienteCierreOut(BaseModel):
    id: str
    incidente_id: str
    resumen_cierre: str
    resultado: str | None = None
    snapshot: dict
    generado_por: UsuarioMini | None = None
    pdf_url: str | None = None
    created_at: datetime
    updated_at: datetime


class ComentarioIncidenteCreateInput(BaseModel):
    contenido: str = Field(min_length=2, max_length=2000)
    es_interno: bool = False


class ZonaCount(BaseModel):
    zona: str
    total: int


class DashboardStats(BaseModel):
    """Métricas agregadas para el dashboard operativo."""

    total: int
    activos: int
    criticos: int
    en_atencion: int
    resueltos_24h: int
    por_zona: list[ZonaCount] = []


class KpiCard(BaseModel):
    """Métrica con valor actual y % de cambio respecto al periodo anterior."""

    valor: float
    cambio_pct: float
    unidad: str = ""


class EvolucionPunto(BaseModel):
    fecha: str  # ISO date YYYY-MM-DD
    total: int
    resueltos: int
    criticos: int


class TipoCount(BaseModel):
    tipo: str
    total: int
    porcentaje: float


class SlaIndicador(BaseModel):
    actual: float
    objetivo: float
    unidad: str  # "min" | "%"


class KpisResponse(BaseModel):
    period: str  # "semana" | "mes" | "trimestre"
    frt: KpiCard
    tmr: KpiCard
    total_incidentes: KpiCard
    tasa_resolucion: KpiCard
    criticos: KpiCard
    sla_cumplimiento: KpiCard
    evolucion: list[EvolucionPunto]
    por_tipo: list[TipoCount]
    por_zona: list[ZonaCount]
    sla: dict[str, SlaIndicador]


class IncidenteDetail(BaseModel):
    id: str
    codigo: str
    titulo: str
    descripcion: str | None = None
    estado: EstadoIncidente
    severidad: NivelSeveridad | None = None
    categoria: str | None = None
    lugar_referencia: str | None = None
    latitud: float | None = None
    longitud: float | None = None
    live_location_enabled: bool = False
    live_location_updated_at: datetime | None = None
    live_location_expires_at: datetime | None = None
    canal_origen: TipoCanal
    fecha_primera_respuesta: datetime | None = None
    fecha_resolucion: datetime | None = None
    created_at: datetime
    updated_at: datetime
    reportante: UsuarioMini | None = None
    operador_asignado: UsuarioMini | None = None
    supervisor: UsuarioMini | None = None
    historial: list[HistorialEvento] = []
    comentarios: list[ComentarioIncidenteItem] = []
    evidencias: list[EvidenciaIncidenteItem] = []
    expediente_cierre: ExpedienteCierreOut | None = None
